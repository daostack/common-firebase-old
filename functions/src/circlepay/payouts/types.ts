import { IBaseEntity } from '../../util/types';

export type PayoutStatus = 'pending';

export interface IPayoutEntity extends IBaseEntity {
  /**
   * This is the id of the payout on circle side. Cam
   * be useful for updating payout status
   */
  circlePayoutId: string;

  /**
   * The ID of the proposal, for witch the payout
   * is made
   */
  proposalId: string;

  /**
   * The destination of the payment
   */
  destination: IPayoutDestination;

  /**
   * The current status of the payout
   */
  status: PayoutStatus;

  /**
   * The amount of the payout in US dollar cents
   */
  amount: number;

  security: IPayoutSecurity[];

  /**
   * Whether the payout has been send to circle for execution
   */
  executed: boolean;

  /**
   * Whether the payout has been voided
   */
  voided: boolean;
}

/**
 * Security details about the payout, needed for the
 * execution of it
 */
export interface IPayoutSecurity {
  /**
   * The ID of the token (no unique globally, but unique in the payout scope)
   */
  id: number;

  /**
   * The token itself
   */
  token: string;

  /**
   * Whether the token has been redeemed successfully
   */
  redeemed: boolean;

  /**
   * How many times an unsuccessful redemption was made
   */
  redemptionAttempts: number;
}

/**
 * Details about the destination
 * of the payout
 */
export interface IPayoutDestination {
  /**
   * The ID of the destination in
   * the firebase
   */
  id: string;

  /**
   * The ID of the destination on
   * circle's side
   */
  circleId: string;
}