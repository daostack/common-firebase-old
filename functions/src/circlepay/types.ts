export interface ICircleCreateCardResponse {
  data: {
    id: string;

    billingDetails: ICircleBillingDetails;

    metadata: {
      email: string;
      phoneNumber: string;
    }

    expMonth: number;
    expYear: number;
    network: 'VISA' | 'MASTERCARD';
    last4: string;
    createDate: Date;
    updateDate: Date;
  }
}



export interface ICircleCreateCardPayload {
  /**
   * Unique idempotency key. This key is utilized to ensure
   * exactly-once execution of mutating requests.
   */
  idempotencyKey: string;

  /**
   * Unique identifier of the public key used in encryption.
   */
  keyId: string;

  /**
   * PGP encrypted json string. The object format given here needs to
   * be stringified and PGP encrypted before it is sent to the server,
   * so encryptedData will end up as a string, rather than an object.
   */
  encryptedData: string;

  billingDetails: ICircleBillingDetails

  /**
   * Two digit number representing the card's expiration month.
   */
  expMonth: number;

  /**
   * Four digit number representing the card's expiration year.
   */
  expYear: number;


  metadata: {
    /**
     * Email of the user
     */
    email: string;

    /**
     * Hash of the session identifier; typically of the end user.
     */
    sessionId: string;

    /**
     * Single IPv4 or IPv6 address of user
     */
    ipAddress: string;
  }
}

interface ICircleBillingDetails {
  /**
   * Full name of the card or bank account holder.
   */
  name: string;

  /**
   * City portion of the address.
   */
  city: string;

  /**
   * Country portion of the address. Formatted as a
   * two-letter country code specified in ISO 3166-1 alpha-2.
   */
  country: string;

  /**
   * Line one of the street address.
   */
  line1: string;

  /**
   * Line two of the street address. Optional
   */
  line2: string;

  /**
   * State / County / Province / Region portion of the address. It is optional
   * except if the country is US or Canada district is required and
   * should use the two-letter code for the subdivision.
   */
  district?: string;

  /**
   * Postal / ZIP code of the address.
   */
  postalCode: string;
}