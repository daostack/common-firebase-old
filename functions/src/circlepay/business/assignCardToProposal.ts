import * as cardDb from '../../util/db/cardDb';
import { NotFoundError } from '../../util/errors';

/**
 * The raw version of assignCard();
 * Please not that here we do not validate anything, so use with caution
 *
 * @param cardId - the id of the card that we want to assign
 * @param proposalId - the id of the proposal that we want to assign to
 *
 * @return { Promise }
 */
export const assignCardToProposal = async (cardId: string, proposalId: string): Promise<void> => {
  const card = (await cardDb.getCardRef(cardId).get()).data();

  if(!card) {
    throw new NotFoundError(cardId, 'card');
  }

  if (card.proposals.some(x => x === proposalId)) {
    // The proposal is already assigned to
    // that card so just return
    return;
  }

  await cardDb.updateCard(cardId, {
    proposals: [
      ...card.proposals,
      proposalId
    ]
  });
};