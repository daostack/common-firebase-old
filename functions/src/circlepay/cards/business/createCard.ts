import axios from 'axios';
import * as yup from 'yup';

import { ErrorCodes } from '../../../constants';
import { circlePayApi } from '../../../settings';
import { validate } from '../../../util/validate';
import { externalRequestExecutor } from '../../../util';
import { billingDetailsValidationSchema } from '../../../util/schemas';

import { ICircleCreateCardPayload, ICircleCreateCardResponse } from '../../types';
import { getCircleHeaders } from '../../index';
import { ICardEntity } from '../types';
import { userDb } from '../../../users/database';
import { cardDb } from '../database';
import { createEvent } from '../../../util/db/eventDbService';
import { EVENT_TYPES } from '../../../event/event';

const createCardValidationSchema = yup.object({
  ownerId: yup.string()
    .required(),

  billingDetails: billingDetailsValidationSchema
    .required(),

  keyId: yup
    .string()
    .required(),

  idempotencyKey: yup
    .string()
    .required(),

  sessionId: yup
    .string()
    .required(),

  ipAddress: yup
    .string()
    .required(),

  encryptedData: yup
    .string()
    .required(),

  expMonth: yup
    .number()
    .min(1)
    .max(12)
    .required(),

  expYear: yup
    .number()
    .min(new Date().getFullYear())
    .max(new Date().getFullYear() + 50)
    .required()
});

type CreateCardPayload = yup.InferType<typeof createCardValidationSchema>;

/**
 * Creates card in the FireStore and on Circle side.
 *
 *
 *
 * @param payload
 */
export const createCard = async (payload: CreateCardPayload): Promise<ICardEntity> => {
  // Validate the passed data
  await validate(payload, createCardValidationSchema);

  // Find the creator
  const user = await userDb.get(payload.ownerId);

  // Format the data for circle
  const headers = await getCircleHeaders();
  const data: ICircleCreateCardPayload = {
    billingDetails: payload.billingDetails as any,
    encryptedData: payload.encryptedData,
    expMonth: payload.expMonth,
    expYear: payload.expYear,
    idempotencyKey: payload.idempotencyKey || payload.sessionId,
    keyId: payload.keyId,
    metadata: {
      email: user.email,
      ipAddress: payload.ipAddress,
      sessionId: payload.sessionId
    }
  };

  // circle card id and user id should be unique :<

  // Create the card on Circle
  const response = await externalRequestExecutor<ICircleCreateCardResponse>(async () => {
    return (await axios.post<ICircleCreateCardResponse>(`${circlePayApi}/cards`,
      data,
      headers
    )).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    userMessage: 'Cannot create the card, because it was rejected by Circle'
  });

  console.log(JSON.stringify(response));

  // If the card was created successfully save it
  const card = await cardDb.add({
    ownerId: user.id,
    circleCardId: response.id,
    metadata: {
      billingDetails: data.billingDetails,
      digits: response.last4,
      network: response.network
    }
  });

  // Create event
  await createEvent({
    type: EVENT_TYPES.
  })

  // Return the created card
  return card;
};