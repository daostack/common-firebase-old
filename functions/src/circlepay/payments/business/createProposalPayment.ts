import { IProposalPayment } from '../types';
import { NotImplementedError } from '../../../util/errors';
import { createPayment } from '../../createPayment';

export const createProposalPayment = async (): Promise<IProposalPayment> => {
  const payment = await createPayment()

  throw new NotImplementedError();
}