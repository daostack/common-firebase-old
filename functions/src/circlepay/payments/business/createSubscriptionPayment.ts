import { ISubscriptionPayment } from '../types';
import { NotImplementedError } from '../../../util/errors';

export const createSubscriptionPayment = async (): Promise<ISubscriptionPayment> => {
  throw new NotImplementedError();
}