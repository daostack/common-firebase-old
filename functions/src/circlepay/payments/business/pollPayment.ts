import { IPaymentEntity } from '../types';
import { ArgumentError } from '../../../util/errors';
import { externalRequestExecutor, poll } from '../../../util';
import { CirclePaymentStatus } from '../../../util/types';
import axios from 'axios';
import { circlePayApi } from '../../../settings';
import { ErrorCodes } from '../../../constants';
import { getCirclePayOptions } from '../../circlepay';
import { ICirclePayment } from '../../types';

interface IPollPaymentOptions {
  interval: number;
  maxRetries: number;

  /**
   * Whether an error will be throw if the payment is finalized, but
   * the status is failed
   */
  throwOnePaymentFailed: boolean;
}

const defaultPaymentOptions: IPollPaymentOptions = {
  interval: 60,
  maxRetries: 32,
  throwOnePaymentFailed: false
};

/**
 * Polls payment until the payment reaches desired state
 *
 * @param payment
 * @param pollPaymentOptions
 */
export const pollPayment = async (payment: IPaymentEntity, pollPaymentOptions?: IPollPaymentOptions): Promise<IPaymentEntity> => {
  if (!payment) {
    throw new ArgumentError('payment', payment);
  }

  const headers = await getCirclePayOptions();
  const options = {
    ...defaultPaymentOptions,
    ...pollPaymentOptions
  };

  const pollFn = async (): Promise<CirclePaymentStatus> => {
    const response = await externalRequestExecutor<ICirclePayment>(async () => {
      return (await axios.get<ICirclePayment>(`${circlePayApi}/payments/${payment.circlePaymentId}`, headers)).data;
    }, {
      errorCode: ErrorCodes.CirclePayError,
      message: 'Polling circle call failed'
    });

    return response.data.status;
  };

  const validateFn = (status: CirclePaymentStatus): boolean => {
    return ['confirmed', 'failed', 'paid'].some(s => s === status);
  };

  const status = await poll<CirclePaymentStatus>(pollFn, validateFn, options.interval, options.maxRetries);

  // @todo Update payment
  // @todo Event
  // @todo Return the updated payment

  return payment;
};