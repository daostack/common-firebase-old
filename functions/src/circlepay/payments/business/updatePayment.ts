import { IPaymentEntity } from '../types';
import { ICirclePayment } from '../../types';
import { failureHelper, feesHelper, isSuccessful } from '../helpers';
import { paymentDb } from '../database';
import { CommonError } from '../../../util/errors';
import { commonDb } from '../../../common/database';
import { proposalDb } from '../../../proposals/database';
import { addCommonMemberByProposalId } from '../../../common/business/addCommonMember';

/**
 * Handles update from circle and saves it to the database
 *
 * @param payment - the current version of the payment from our FireStore
 * @param circlePayment - the current version of the payment as is from Circle
 */
export const updatePayment = async (payment: IPaymentEntity, circlePayment: ICirclePayment): Promise<IPaymentEntity> => {
  let updatedPayment: IPaymentEntity = payment;

  switch (circlePayment.data.status) {
    case 'failed':
      updatedPayment = {
        ...payment,

        status: circlePayment.data.status,
        failure: failureHelper.processFailedPayment(circlePayment)
      };

      break;
    case 'confirmed':
    case 'paid':
      updatedPayment = {
        ...payment,

        status: circlePayment.data.status,
        fees: feesHelper.processCircleFee(circlePayment)
      };

      break;
    default:
      logger.warn('Unknown payment state occurred. Not knowing how to handle the payment update', {
        payment,
        circlePayment: circlePayment.data,
        unknownStatus: circlePayment.data.status
      });

      break;
  }

  updatedPayment = await paymentDb.update(updatedPayment);

  // If the status has change broadcast an event
  if (payment.status !== updatedPayment.status) {
    logger.debug('Payment status update occurred on payment', {
      status: `${payment.status} -> ${updatedPayment.status}`,
      paymentType: payment.type,
      paymentId: payment.id
    }, {
      oldPayment: payment,
      updatedPayment: updatedPayment
    });

    // Create the payment updated event
    switch (updatedPayment.status) {
      case 'failed':
      case 'confirmed':
      case 'paid':
      case 'pending':
        // @todo Implement the payment status changed events
        // logger.error('NotImplementedError: Payment status changes events');

        break;
      default:
        throw new CommonError(`The payment status has updated, but is not known.`, {
          payment: updatedPayment
        });
    }

    // @todo If this is subscription payment handle subscription update
    if (payment.type === 'subscription') {
      logger.warn('Not Implemented Waring: Payment status change for subscription payment');

      // Handle this only if the previous status was explicitly pending
      // and the new one is successful
      // if (payment.status === 'pending' && isSuccessful(payment)) {
      //
      // }
    }

    // @todo If this is proposal payment handle the proposal update
    if (payment.type === 'one-time') {
      logger.warn('Not Implemented Waring: Payment status change for subscription payment');

      // Handle this only if the previous status was explicitly pending
      // and the new one is successful
      if (payment.status === 'pending' && isSuccessful(payment)) {
        logger.notice('Updating data for hanging payment', {
          paymentId: payment.id
        });

        const proposal = await proposalDb.getJoinRequest(payment.proposalId);

        // Update common funding info
        const common = await commonDb.get(proposal.commonId);

        common.raised += proposal.join.funding;
        common.balance += proposal.join.funding;

        await commonDb.update(common);

        // Add the user as member
        await addCommonMemberByProposalId(proposal.id);
      }
    }
  }

  return updatedPayment;
};