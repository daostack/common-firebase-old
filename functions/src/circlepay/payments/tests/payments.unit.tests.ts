import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import {
  getPaymentFailedResponse,
  getPaymentPendingResponce,
  getPaymentSuccessfulResponce
} from './data/testCircleResponses';
import { ICirclePayment } from '../../types';
import '../../../util/tests/helpers/mockers/payments/updatePayment.mocker';
import '../../../util/tests/helpers/mockers/firebase.mocker';

// Add the logger to the tests globals
import '../../../util/logger';

import { pendingPaymentEntity, successfulPaymentEntity } from './data/testPaymentEntities';
import { pollPayment } from '../business/pollPayment';

const circleMatcher = /.+circle\.com\/.*/;
const circePaymentEndpointMatcher = new RegExp(circleMatcher.source + (/\/payments/).source);

// ----- Local Mocks

jest.mock('firebase-admin', () => ({
  initializeApp: jest.fn(),
  firestore: () => ({
    collection: jest.fn(() => ({
      withConverter: jest.fn()
        .mockImplementation(() => jest.fn())
      //
      // where: jest.fn(queryString => ({
      //   get: mockQueryResponse
      //
      // }))
    })),
    settings: jest.fn()
  })
}));

describe('Payment Unit Tests', () => {
  describe('Payment polling', () => {
    it('should poll until successful status is reached', async () => {
      // Arrange
      const mock = new MockAdapter(axios);
      const mockSpy = jest.spyOn(axios, 'get');


      // On the initial 10 calls return pending
      for (let i = 0; i < 10; i++) {
        mock
          .onGet(circePaymentEndpointMatcher)
          .replyOnce<ICirclePayment>(200, getPaymentPendingResponce);
      }

      // On the 11th and all sequential calls return check success
      mock
        .onGet(circePaymentEndpointMatcher)
        .reply<ICirclePayment>(200, getPaymentSuccessfulResponce);

      // Act
      const pollResult = await pollPayment(successfulPaymentEntity, {
        interval: 0
      });

      // Assert
      expect(pollResult).toMatchSnapshot();
      expect(mockSpy).toHaveBeenCalledTimes(11);

      // Cleanup
      mock.restore();
      mockSpy.mockClear();
    });

    it('should poll until failed status is reached', async () => {
      // Arrange
      const mock = new MockAdapter(axios);
      const mockSpy = jest.spyOn(axios, 'get');


      // On the initial 10 calls return pending
      for (let i = 0; i < 10; i++) {
        mock
          .onGet(circePaymentEndpointMatcher)
          .replyOnce<ICirclePayment>(200, getPaymentPendingResponce);
      }

      // On the 11th and all sequential calls return check success
      mock
        .onGet(circePaymentEndpointMatcher)
        .reply<ICirclePayment>(200, getPaymentFailedResponse);

      // Act
      const pollResult = await pollPayment(pendingPaymentEntity, {
        interval: 0
      });

      // Assert
      expect(pollResult).toMatchSnapshot();
      expect(mockSpy).toHaveBeenCalledTimes(11);

      // Cleanup
      mock.restore();
      mockSpy.mockClear();
    });

    it('should throw if failed status is reached and throw was requested', async () => {
      // Arrange
      const mock = new MockAdapter(axios);
      const mockSpy = jest.spyOn(axios, 'get');

      mock
        .onGet(circePaymentEndpointMatcher)
        .reply<ICirclePayment>(200, getPaymentFailedResponse);

      // Act and Assert
      await expect(pollPayment(pendingPaymentEntity, {
        interval: 0,
        throwOnPaymentFailed: true
      }))
        .rejects
        .toThrowError('Payment failed');

      // Cleanup
      mock.restore();
      mockSpy.mockClear();
    });

    it('should throw if the maximum poll attempts are reached', async () => {
      // Arrange
      const mock = new MockAdapter(axios);
      const mockSpy = jest.spyOn(axios, 'get');

      // Only pending responses
      mock
        .onGet(circePaymentEndpointMatcher)
        .reply<ICirclePayment>(200, getPaymentPendingResponce);

      // Act and assert
      await expect(pollPayment(pendingPaymentEntity, {
        interval: 0
      }))
        .rejects
        .toThrowError('Max polling attempts reached!');

      // Cleanup
      mock.restore();
      mockSpy.mockClear();
    });
  });

  // describe('General payments', () => {
  //
  // });
  //
  // describe('One-time payments', () => {
  //
  // });
  //
  // describe('Subscription payments', () => {
  //
  // });
});