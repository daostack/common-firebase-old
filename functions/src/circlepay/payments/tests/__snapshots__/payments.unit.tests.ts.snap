// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment Unit Tests General payments should successfully create payment given valid data 1`] = `
Object {
  "amount": Object {
    "amount": 10000,
    "currency": "USD",
  },
  "circlePaymentId": "62d74450-9b61-4d5a-8890-4fb22eb09fd8",
  "createdAt": Object {
    "nanoseconds": 0,
    "seconds": 1608168243,
  },
  "fees": Object {
    "amount": 0,
    "currency": "USD",
  },
  "id": "ae6b689c-5dcb-41ed-84a7-51e6ce7faa76",
  "proposalId": "ae6b689c-5dcb-41ed-84a7-51e6ce7faa76",
  "source": Object {
    "id": "ae6b689c-5dcb-41ed-84a7-51e6ce7faa76",
    "type": "card",
  },
  "status": "pending",
  "subscriptionId": undefined,
  "type": "one-time",
  "updatedAt": Object {
    "nanoseconds": 0,
    "seconds": 1608176083,
  },
  "userId": "ae6b689c-5dcb-41ed-84a7-51e6ce7faa76",
}
`;

exports[`Payment Unit Tests One-time payments should create the payment given the correct payload 1`] = `
Object {
  "amount": 1598,
  "currency": "USD",
}
`;

exports[`Payment Unit Tests Payment polling should poll until failed status is reached 1`] = `
Object {
  "amount": Object {
    "amount": 36900,
    "currency": "USD",
  },
  "circlePaymentId": "b3025713-7988-45e5-8af8-cca8d624ce28",
  "createdAt": Object {
    "nanoseconds": 0,
    "seconds": 1608168243,
  },
  "failure": Object {
    "errorCode": "payment_failed",
    "errorDescription": "Payment failed due to unspecified error",
  },
  "fees": Object {
    "amount": 1598,
    "currency": "USD",
  },
  "id": "0ef989c3-d096-42c1-b525-66a3719e413a",
  "proposalId": "706da88a-a417-464d-8d36-3d68fb2b954c",
  "source": Object {
    "id": "62e367b3-6e16-44d8-9407-d5f9486924f4",
    "type": "card",
  },
  "status": "failed",
  "subscriptionId": "811dd5ed-1481-43bc-bc79-cacf6598916f",
  "type": "subscription",
  "updatedAt": Object {
    "nanoseconds": 0,
    "seconds": 1608176083,
  },
  "userId": "2Ti3LkP23KUJVp2AZY3wVHYWRxm2",
}
`;

exports[`Payment Unit Tests Payment polling should poll until successful status is reached 1`] = `
Object {
  "amount": Object {
    "amount": 36900,
    "currency": "USD",
  },
  "circlePaymentId": "b3025713-7988-45e5-8af8-cca8d624ce28",
  "createdAt": Object {
    "nanoseconds": 0,
    "seconds": 1608168243,
  },
  "fees": Object {
    "amount": 1598,
    "currency": "USD",
  },
  "id": "0ef989c3-d096-42c1-b525-66a3719e413a",
  "proposalId": "706da88a-a417-464d-8d36-3d68fb2b954c",
  "source": Object {
    "id": "62e367b3-6e16-44d8-9407-d5f9486924f4",
    "type": "card",
  },
  "status": "paid",
  "subscriptionId": "811dd5ed-1481-43bc-bc79-cacf6598916f",
  "type": "subscription",
  "updatedAt": Object {
    "nanoseconds": 0,
    "seconds": 1608176083,
  },
  "userId": "2Ti3LkP23KUJVp2AZY3wVHYWRxm2",
}
`;
